sale_order_product_recommendation/wizards/sale_order_recommendation.py

```diff
@@ -46,6 +46,23 @@ class SaleOrderRecommendation(models.TransientModel):
     def _default_order_id(self):
         return self.env.context.get("active_id", False)
 
+    def _recomendable_sale_order_lines_domain(self):
+        """Domain to find recent SO lines."""
+        start = datetime.now() - timedelta(days=self.months * 30)
+        start = fields.Datetime.to_string(start)
+        other_sales = self.env["sale.order"].search([
+            ("partner_id", "child_of",
+             self.order_id.partner_id.commercial_partner_id.id),
+            ("date_order", ">=", start),
+        ])
+        return [
+            ("order_id", "!=", self.order_id.id),
+            ("order_id", "in", other_sales.ids),
+            ("product_id.active", "=", True),
```

Comment:
As most of the products will be active and probably `sale_ok`, I would bet for doing this in a later filter, but I'm not sure if you have already tried this and measure performance

Reply:
Given we have to ask the DB for a recordset, it's better to filter it directly in the DB. It's always faster than downloading extra recordsets just to ignore them later in Python code.

If the recordsets were already cached, that could be a different story, but it's not the case here.

Reply:
The problem is that this is not auto-joined, so it actually requires several queries + the problem with `in` operator in SQL a list with thousand of elements.

Reply:
I have been profiling this.

I didn't understand what you were saying, but now I see that indeed these filters are made in separate SQL queries. That's quite strange, since it could just be added to the same query. I guess that's to apply permissions...

Anyways, removing these leafs from the domain we save 2 queries and ~7ms in total, but later we'd have to browse all product records and manually prefetch their `active` and `sale_ok` fields to actually end up saving 1 query, at the cost of code readability.

IMHO it's not worth it.

Reply:
We would save that queries if `product_id` has `auto_join=True`, but that disables ACL check. In this case, maybe it makes sense to add it. In current state, the problem is not the extra 2 queries, but having thousand of product, creating the third query: `('product_id', 'in', [<thousand of ids>])`, which the slowest one. Have you tried?

Reply:
I haven't tried that, but I doubt that is a performance problem.

I think you are in a confusion with the performance problems we have faced in the past with subqueries, not for lists of ids. I mean, it's more things like `where id in (select id from lalala)`, because if the server has not enough worker memory, it executes the subquery record by record and the execution time grows exponentially. But that's a completely different case.

Even in Python, if you use the `in` operator with a `set()`, you get instant results no matter the set size. I imagine Postgres uses a similar optimization under the hood to avoid that problem.

I'd say the best we can do is merge and then patch it if we really see some real bottleneck in production. Otherwise we might be wasting time in fixing something that is not broken.

Reply:
Seen directly in a meeting and it doesn't seem to affect the performance, so we keep it this way.